src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,688 @@
+                    actionsBuilder.Add(new MyCodeAction(CreateDisplayText(allOccurrences: true, trampoline: true, overload: false),
+                        c => IntroduceParameterAsync(semanticDocument, expression, true, true, false, cancellationToken)));
+                    actionsBuilder.Add(new MyCodeAction(CreateDisplayText(allOccurrences: false, trampoline: false, overload: true),
+                        c => IntroduceParameterAsync(semanticDocument, expression, false, false, true, cancellationToken)));
+                    actionsBuilder.Add(new MyCodeAction(CreateDisplayText(allOccurrences: true, trampoline: false, overload: true),
+                        c => IntroduceParameterAsync(semanticDocument, expression, true, false, true, cancellationToken)));
+                }
+            }
+
+            return actionsBuilder.ToImmutable();
+        }
+
+        private static string CreateDisplayText(bool allOccurrences, bool trampoline, bool overload)
+                => (allOccurrences, trampoline, overload) switch
+                {
+                    (true, true, false) => FeaturesResources.Extract_method_to_invoke_at_all_callsites_for_all_occurrences_of_0,
+                    (true, false, false) => FeaturesResources.Introduce_parameter_for_all_occurrences_of_0,
+                    (true, false, true) => FeaturesResources.Introduce_overload_with_new_parameter_for_all_occurrences_of_0,
+                    (false, true, false) => FeaturesResources.Extract_method_to_invoke_at_all_callsites_for_0,
+                    (false, false, true) => FeaturesResources.Introduce_overload_with_new_parameter_for_0,
+                    (false, false, false) => FeaturesResources.Introduce_parameter_for_0,
+                    _ => throw new System.NotImplementedException()
+                };
+
+        /// <summary>
+        /// Finds the matches of the expression within the same block
+        /// </summary>
+        protected static ISet<TExpressionSyntax> FindMatches(SemanticDocument originalDocument,
+            TExpressionSyntax expressionInOriginal, SemanticDocument currentDocument, SyntaxNode withinNodeInCurrent,
+            bool allOccurrences, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = currentDocument.Document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var originalSemanticModel = originalDocument.SemanticModel;
+            var currentSemanticModel = currentDocument.SemanticModel;
+
+            var result = new HashSet<TExpressionSyntax>();
+            var matches = from nodeInCurrent in withinNodeInCurrent.DescendantNodesAndSelf().OfType<TExpressionSyntax>()
+                          where NodeMatchesExpression(originalSemanticModel, currentSemanticModel, expressionInOriginal, nodeInCurrent, allOccurrences, cancellationToken)
+                          select nodeInCurrent;
+            result.AddRange(matches.OfType<TExpressionSyntax>());
```

Comment:
why do we need old and new semantic models here?

Reply:
The SemanticEquivalance requires a "new" and "old", I just pass in the same one bc there's no distinction. I'll just pass in one and reuse it instead